src/core/lib/event_engine/grpc_ares_wrapper.cc

```diff
@@ -0,0 +1,1013 @@
+          GRPC_ARES_WRAPPER_TRACE_LOG("request:%p notify write on: %d", this,
+                                      fd_node->as);
+          fd_node->writable_registered = true;
+          fd_node->polled_fd->RegisterForOnWriteableLocked(
+              [self = Ref(DEBUG_LOCATION, "Work"),
+               fd_node](absl::Status status) mutable {
+                grpc_core::ApplicationCallbackExecCtx callback_exec_ctx;
+                grpc_core::ExecCtx exec_ctx;
+                self->OnWritable(fd_node, status);
+                self.reset();
+              });
+        }
+      }
+    }
+  }
+  // Any remaining fds in ev_driver->fds were not returned by ares_getsock()
+  // and are therefore no longer in use, so they can be shut down and removed
+  // from the list.
+  while (!fd_node_list_->IsEmpty()) {
+    FdNode* fd_node = fd_node_list_->PopFdNode();
+    if (!fd_node->already_shutdown) {
+      GRPC_ARES_WRAPPER_TRACE_LOG("request: %p shutdown fd: %s", this,
+                                  fd_node->polled_fd->GetName());
+      fd_node->polled_fd->ShutdownLocked(absl::OkStatus());
+      fd_node->already_shutdown = true;
+    }
+    if (!fd_node->readable_registered && !fd_node->writable_registered) {
+      GRPC_ARES_WRAPPER_TRACE_LOG("request: %p delete fd: %s", this,
+                                  fd_node->polled_fd->GetName());
+      delete fd_node;
+    } else {
+      new_list->PushFdNode(fd_node);
+    }
+  }
+  fd_node_list_.swap(new_list);
+}
+
+void GrpcAresRequest::StartTimers() {
+  // Initialize overall DNS resolution timeout alarm
+  EventEngine::Duration timeout =
+      timeout_.count() == 0 ? EventEngine::Duration::max() : timeout_;
+  GRPC_ARES_WRAPPER_TRACE_LOG("request:%p StartTimers timeout in %" PRId64
+                              " ms",
+                              this, Milliseconds(timeout));
+
+  Ref(DEBUG_LOCATION, "StartTimers").release();
+  query_timeout_handle_ = event_engine_->RunAfter(timeout, [this] {
+    grpc_core::ApplicationCallbackExecCtx callback_exec_ctx;
+    grpc_core::ExecCtx exec_ctx;
+    OnQueryTimeout();
+  });
+
+  // Initialize the backup poll alarm
+  EventEngine::Duration next_ares_backup_poll_alarm_duration =
+      calculate_next_ares_backup_poll_alarm_duration();
+  GRPC_ARES_WRAPPER_TRACE_LOG(
+      "request:%p StartTimers next ares process poll time in %" PRId64 " ms",
+      this, Milliseconds(next_ares_backup_poll_alarm_duration));
+
+  Ref(DEBUG_LOCATION, "StartTimers").release();
+  ares_backup_poll_alarm_handle_ =
+      event_engine_->RunAfter(next_ares_backup_poll_alarm_duration, [this] {
+        grpc_core::ApplicationCallbackExecCtx callback_exec_ctx;
+        grpc_core::ExecCtx exec_ctx;
+        OnAresBackupPollAlarm();
+      });
+}
+
+void GrpcAresRequest::CancelTimers() {
+  if (query_timeout_handle_.has_value()) {
+    if (event_engine_->Cancel(*query_timeout_handle_)) {
+      Unref(DEBUG_LOCATION, "CancelTimers");
```

Comment:
if the timer callback owns the ref, then if cancellation succeeds, the callback will be deleted and the RefCountedPtr it owns will subtract one when destroyed.